dashboard/app/models/user.rb

```diff
@@ -502,8 +522,18 @@ def user_progress_by_stage(stage)
   end
 
   def user_level_for(script_level, level)
+    # see if we have cached all user levels for this script
+    if user_levels_cached_for_script?(script_level.script_id)
+      # if so, return from cache. Note that nil is a possible return value from cache
+      # (which means specified saved user level does not exist). We deliberately don't fall through
+      # to a database dip if nil.
+      return user_level_from_cache(script_level.script_id, level.id)
+    end
+
+    # Do a database query to get the user_level. We deliberately do not try to further cache this
+    # result.
     user_levels.find_by(script_id: script_level.script_id,
-                        level_id: level.id)
+      level_id: level.id)
```

Comment:
nit: If you are adjusting the whitespace, any reason not to combine onto previous line?